Add event message

MoonBow-1 opened the pull request:
A smaller PR from #15, which brings the EventMessage related classes.

- Tested using the JMeter test plan
- Does not fix the setters that are used in `HttpEventData`

MoonBow-1 committed:
Cherry-pick EventMessage objects fron fix-mocking branch

MoonBow-1 committed:
Use EventMessage in EventManager

MoonBow-1 committed:
Cherry-pick pom.xml and cfe-16-test-plan.jmx from fix-mocking branch

MoonBow-1 committed:
Make variables final in EventMessageImpl

MoonBow-1 committed:
Make variables final in EventMessageStub

MoonBow-1 committed:
Make variables final in JsonEventImpl

MoonBow-1 committed:
Fix unit test names and add one test for JsonEventImpl

MoonBow-1 committed:
Fix test server port in ServiceAndEventManagerIT

MoonBow-1 committed:
Remove redundant code from EventManager.convertData

MoonBow-1 committed:
Fix broken test due to changed test server port

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventMessageStub.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public int hashCode() {
+        throw new UnsupportedOperationException("EventStub does not support this");
```
stub is comparable to another stub, please implement the hashCode method

MoonBow-1 replied:
Fixed in commit b021310e

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventMessageStub.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public boolean equals(final Object obj) {
+        throw new UnsupportedOperationException("EventStub does not support this");
```
stub is comparable to another stub, please implement the equals method

MoonBow-1 replied:
Fixed in commit b021310e

kortemik commented on src/test/java/com/teragrep/cfe_16/event/EventMessageImplTest.java:
```diff
@@ -0,0 +1,87 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class EventMessageImplTest {
```
equalsverifier testcase not present, please add

MoonBow-1 replied:
EqualsVerifier test implemented in commit 86b339bc

kortemik commented on src/test/java/com/teragrep/cfe_16/event/EventMessageStubTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    @DisplayName("hashCode() throws UnsupportedOperationException if called")
```
hashCode must return a valid hashCode that is comparable and also same for the stub instances created with same content meaning usually all as stubs do not encapsulate things

MoonBow-1 replied:
Fixed in commit b021310e

kortemik commented on src/test/java/com/teragrep/cfe_16/event/EventMessageStubTest.java:
```diff
@@ -0,0 +1,94 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class EventMessageStubTest {
```
equalsverifier testcase mssing

MoonBow-1 replied:
EqualsVerifier added and test implemented in commit 76a1a021

kortemik commented on src/test/java/com/teragrep/cfe_16/event/EventMessageStubTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    @DisplayName("equals() throws UnsupportedOperationException if called")
```
equals must produce a proper equals check

MoonBow-1 replied:
Fixed in commit https://github.com/teragrep/cfe_16/commit/b021310ec84f83958327d006208b225892c07bf7

kortemik commented on src/main/java/com/teragrep/cfe_16/EventManager.java:
```diff
@@ -142,30 +143,24 @@ public ObjectNode convertData(
+                syslogMessages.add(converter.httpToSyslog(eventData));
+            }
+            catch (final JsonProcessingException e) {
+                LOGGER.error(e.getMessage());
```
only try, catch, log, rethrow is permitted. this does not rethrow, is there a reason?

MoonBow-1 replied:
In case the ObjectMapper throws an JsonProcessingException, should we stop processing the current batch of events, and return an error to the user to indicate failure?

Currently, if the batch completes in `EventManager`:
```java
ObjectNode responseNode = this.objectMapper.createObjectNode();
responseNode.put("text", "Success");
responseNode.put("code", 0);
```

This ObjectNode is returned in `HECServiceImpl`:
```java
ObjectNode ackNode = this.eventManager
                .convertData(authToken, channel, eventInJson, headerInfo, this.acknowledgements);
return ackNode;
```

That return value is used in `HECRestController`, where timing are calculated, before finally returning the Node to the user:
```java
@Autowired
private HECService service;
...
JsonNode response = service.sendEvents(request, channel, eventInJson);
```
---

In case JsonProcessingException is thrown:
1. Change method signature of `service.sendEvents(...)` to throw an `JsonProcessingException`
2. Catch that Exception in `HECRestController`
3. Stop processing the request ?
4. Log the Exception ?
5. Instead of returning the JsonNode as a response, create a new Node which informs the user of a failure ?
```java
try {
    final JsonNode response = service.sendEvents(request, channel, eventInJson);
}
catch (final JsonProcessingException e) {
    LOGGER
            .error(
                    "EventManager threw an Exception when parsing the JSON events with message <{}>",
                    e.getMessage()
            );
    return new ObjectMapper().createObjectNode().put("text", "Failure");
}
...
```

MoonBow-1 replied:
Implemented in commit 58aeda00:

1. `EventManager.convertData` now throws `JsonProcessingException` or `JsonSyntaxException` if parsing an event fails
2. These Exceptions are caught in `HECServiceImpl`
3. Request metadata (headers, host, user agent etc.) is gathered
4. That metadata is logged to console alongside the actual exception that happened
5. `ExceptionJsonResponse` is returned to `HECRestController`
6. Timings are still calculated
7. `ExceptionJsonResponse` is returned to used with `.asJsonString()`

---

`HECServiceImpl`:
```java
try {
    return this.eventManager.convertData(authToken, channel, eventInJson, headerInfo, this.acknowledgements);
}
catch (final JsonProcessingException | JsonSyntaxException e) {
    final ExceptionEventContext exceptionEventContext = new ExceptionEventContext(
            headerInfo,
            request.getHeader("user-agent"),
            request.getRequestURI(),
            request.getRemoteHost()
    );
    final ExceptionEvent event = new ExceptionEvent(exceptionEventContext, UUID.randomUUID(), e);
    event.logException();
    return new ExceptionJsonResponse(HttpStatus.BAD_REQUEST, event);
}
```

Example of a logged exception in console:
```
[ERROR]  2025-09-10 15:03:49  :  Technical error while processing request <[ExceptionEventContext{headerInfo=HeaderInfo{xForwardedFor=XForwardedForImpl{headerValue='example-For'}, xForwardedHost=XForwardedHostImpl{headerValue='example-Host'}, xForwardedProto=XForwardedProtoImpl{headerValue='example-Proto'}}, userAgent='Apache-HttpClient/4.5.14 (Java/11.0.25)', uriPath='/services/collector', host='127.0.0.1'}]> correlation id <0f134e13-3a3b-4181-ae93-44f1923c70db>
com.google.gson.JsonSyntaxException: java.io.EOFException: End of input at line 2 column 1 path $.event
	at com.google.gson.internal.Streams.parse(Streams.java:59) ~[gson-2.10.1.jar:?]
	at com.google.gson.JsonStreamParser.next(JsonStreamParser.java:89) ~[gson-2.10.1.jar:?]
	at com.teragrep.cfe_16.EventManager.convertData(EventManager.java:157) ~[classes/:?]
	at com.teragrep.cfe_16.service.HECServiceImpl.sendEvents(HECServiceImpl.java:185) [classes/:?]
	at com.teragrep.cfe_16.rest.HECRestController.sendEvents(HECRestController.java:117) [classes/:?]
...
```

---

`ExceptionJsonResponse`:
```java
  public String asJsonString() {
    final ObjectMapper jsonObjectBuilder = new ObjectMapper();
    return jsonObjectBuilder
            .createObjectNode()
            .put(
                    "message",
                    "An error occurred while processing your Request. See event id in the technical log for details."
            )
            .put("id", exceptionEvent.uuid().toString())
            .toString();
}
  ```

kortemik reviewed (changes requested):
will re-review once comments are fixed

MoonBow-1 committed:
Fix hashCode and equals methods for EventMessageStub

MoonBow-1 committed:
Add equalsVerifier test to EventMessageStubTest

- Adds equalsVerifier to project dependencies

MoonBow-1 committed:
Fix EventMessageImpl.equals method's logic and test it with equalsVerifier

MoonBow-1 committed:
Initial Response classes and tests for them

Laukkala commented on src/main/java/com/teragrep/cfe_16/response/Response.java:
```diff
@@ -0,0 +1,58 @@
+
+    public abstract HttpStatus status();
+
+    public abstract ObjectNode body();
```
Perhaps body() could return a String instead? It seems that calls to body() are often appended with .toString() anyway.

Alternatively, this Interface could be renamed to JsonResponse, as the body is provided as a jackson ObjectNode, which seems to be representing JSON data. Of course renaming this would also require renaming the existing concrete class JsonResponse to something else, perhaps a StandardJsonResponse

Laukkala commented on src/main/java/com/teragrep/cfe_16/response/ExceptionJsonResponse.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    public ObjectNode body() {
+        LOGGER.error("Event_{}", eventId, throwable);
```
Exception gets logged only after body() is called. This could lead to trouble if a Response is generated, but the developer doesn't read the body of the Response or if the body() call is not reached.
```
Response response = service.sendEvents(request, channel, eventInJson); // <--- Exception here  
// Other code is executed. 
if( response.status != HttpStatus.OK_200 ) {
    response.body(); // <--- Exception gets printed to logs here
} 
```

A workaround might be to create another object (such as a ExceptionEvent) that is responsible for generating an UUID and printing the Exception (with the UUID) to logs, and that object being used as the constructor parameter for ExceptionJsonResponse.

Then ExceptionJsonResponse will generate an error message for the user and telling which event ID they should look for in the logs.

This way you can log the Exception first, and generate the response later.

MoonBow-1 replied:
Did you have something like this in mind?

EventManager's `convertData(...)`:
```java
catch (final JsonProcessingException e) {
  return new ExceptionJsonResponse(
          HttpStatus.BAD_REQUEST,
          e,
          new ExceptionEvent(UUID.randomUUID()).exceptionMessage()
  );
}
```

with `ExceptionEvent` logging the event when the exceptionMessage is retrieved?
```java
private static final Logger LOGGER = LoggerFactory.getLogger(ExceptionEvent.class);
private final UUID uuid;
...
public String exceptionMessage() {
    LOGGER.error("Event_{}", uuid);
    return "An error occurred while processing your Request. See event id " + uuid
            + " in the technical log for details.";
}
```

This leaves the `ExceptionJsonResponse` responsible for constructing the `ObjectNode` (so far, I believe a "JsonString" method works here, so the `ObjectNode` is made into a String in the `body()` method, which can be returned to the user.
```java
private final String message;
private final HttpStatus status;
private final Throwable throwable;
...
public ObjectNode body() {
    final ObjectMapper jsonObjectBuilder = new ObjectMapper();
    return jsonObjectBuilder.createObjectNode().put("message", message);
}
```

Laukkala replied:
I thought about it this way:

`EventManager.convertData()`
Here the Exception is caught, an UUID is generated for it, and we call logExceptionEvent() to print the Exception into the logs.
After logging, we generate an ExceptionJsonResponse, and return it to the caller.

```java
catch (final JsonProcessingException e) {
  ExceptionEvent event = new ExceptionEvent(UUID.randomUUID(),e);
  event.logExceptionEvent();
  return new ExceptionJsonResponse(
          HttpStatus.BAD_REQUEST,
          event
  );
}
```

`ExceptionEvent`
Responsible for printing the Exception and its ID to the logs. It can return it's ID when asked.
```java
private static final Logger LOGGER = LoggerFactory.getLogger(ExceptionEvent.class);
private final UUID uuid;
private final Throwable throwable;
...
public void logExceptionEvent() {
    LOGGER.error("Event_"+uuid, throwable);
}

public UUID uuid(){
    return uuid;
}
```

`ExceptionJsonResponse`
Responsible for generating a message to be sent back to the end user.
This will take an ExceptionEvent, from which it can query the ID of the event.
```java
private final ExceptionEvent exceptionEvent;
private final HttpStatus status;
...
public ObjectNode body() {
    final ObjectMapper jsonObjectBuilder = new ObjectMapper();
    return jsonObjectBuilder.createObjectNode().put("message", "An error occurred while processing your Request. See event id " + exceptionEvent.uuid()
            + " in the technical log for details.";);
}
```

It would be nice if ExceptionEvent logged the Exception as soon as it is instantiated, so you wouldn't have to explicitly call .logExceptionEvent(), but I'm not sure how to do that without adding code in the constructor

MoonBow-1 replied:
This has been implemented in commit https://github.com/teragrep/cfe_16/pull/47/commits/9246dd0f4fb627a3d8018a0d2e8751c9a68bbb13

Laukkala replied:
Now Exceptions are logged immediately, and the response also includes the event ID, so this looks resolved to me!

Laukkala reviewed (changes requested):
Only real issue I found was with the fact that Exceptions might not get logged if Response.body() is not called for some reason. Currently it seems that body() is being called eventually, but it might cause problems later on if another developer forgets to call it after receiving a Response. I would suggest separating the responsibility of writing to logs to another object, and leave the responsibility of generating the user-facing message as it is.

MoonBow-1 committed:
Implement ExceptionEvent class and print the Exception to the logs already in EventManager

- Ensures that the log is printed, even before the response is sent or (mis)handled

Laukkala reviewed (approved):
With the changes to logging, this looks good to me once review comments from other reviewers have been resolved

MoonBow-1 committed:
Assert that responses are HTTP code 400 and message is "Success"

MoonBow-1 committed:
Change Response.body() to returns a String, and rename it to Response.asJsonString

MoonBow-1 committed:
Fix Jmeter success assertion due to wrong status code

kortemik commented on src/main/java/com/teragrep/cfe_16/response/ExceptionJsonResponse.java:
```diff
@@ -0,0 +1,97 @@
+                .createObjectNode()
+                .put(
+                        "message",
+                        "An error occurred while processing your Request. See event id " + exceptionEvent.uuid()
```
Change to "Please contact administrator of the service and provied event id for investigation."

MoonBow-1 replied:
Should we provide the event ID in another field or provide it as
`
"Please contact administrator of the service and provide event id {uuid} for investigation."
`

kortemik replied:
If new field is easy to add and is still readable that might make sense so the message does not need to be parsed, in case using machine reading to errors.

MoonBow-1 replied:
Implementation done in commit 2fd3354f

kortemik commented on src/main/java/com/teragrep/cfe_16/response/ExceptionEvent.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    public void logException() {
+        LOGGER.error("Event_{}", uuid, throwable);
```
Logging convention does not match "Procedure: Variable Logging" please change to
LOGGER.error("Technical error while processing request <[{}]> correlation id <{}>", request, uuid, throwable);

MoonBow-1 replied:
Fixed in commit e245e1b3

kortemik replied:
fix is lacking the request part

MoonBow-1 replied:
Request added to ExceptionEvent in commit 16ba66de

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEventImpl.java:
```diff
@@ -0,0 +1,108 @@
+        if (this.jsonNode != null && this.jsonNode.isObject()) {
+            return this.jsonNode;
+        }
+        throw new IllegalStateException("jsonEvent node not valid");
```
IllegalArgumentException as jsonNode from constructor argument is bad

MoonBow-1 replied:
IllegalArgumentException now thrown in commit f5e22bef

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEventImpl.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Override
+    public EventMessage asEvent() {
```
multiple return statements cause branching, please refactor to single point of return. @q22u please update documentation about branching to contain single point of return guidance

MoonBow-1 replied:
Fixed in commit bd31ad13

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEvent.java:
```diff
@@ -43,13 +43,15 @@
-    @Test
-    public void contextLoads() {
-    }
+    EventMessage asEvent();
```
asEventMessage is a preferred naming as it returns such

MoonBow-1 replied:
Fixed in commit ed4e75f1

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEvent.java:
```diff
@@ -43,13 +43,15 @@
-    }
+    EventMessage asEvent();
+
+    JsonNode asNode();
```
is this payloadJsonNode()?

MoonBow-1 replied:
Renamed in commit a0755f4a

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEvent.java:
```diff
@@ -43,13 +43,15 @@
+
+    JsonNode asNode();
+
+    JsonNode asTimeNode();
```
is this timeJsonNode()?

MoonBow-1 replied:
Renamed in commit 0d33f21c

kortemik commented on src/main/java/com/teragrep/cfe_16/EventManager.java:
```diff
@@ -142,30 +148,26 @@ public ObjectNode convertData(
-         * successful.
-         */
-        previousEvent = null;
+                syslogMessages.add(converter.httpToSyslog(eventData));
```
please do not nest multiple method calls. this will cause stacktraces to be hard to read as it only says exception at L163 and not that if .add or .httpToSyslog threw. @q22u please add instructions to documentation about not to nest method calls.

MoonBow-1 replied:
Chained method calls removed in commit 11786218

kortemik commented on src/main/java/com/teragrep/cfe_16/rest/HECRestController.java:
```diff
@@ -79,47 +80,48 @@ public class HECRestController {
         // TODO: Try to think an alternative way to implement getting the body of the
         // call
-        String eventInJson = requestBodyCleaner.cleanAckRequestBody(body.toString(), channel);
+        final String eventInJson = requestBodyCleaner.cleanAckRequestBody(body.toString(), channel);
```
please create an issue about refactoring RequestBodyCleaner out from the code. it seems to do some kind of url parameter key parsing with a strange approach.

MoonBow-1 replied:
Issue #48 created

MoonBow-1 committed:
Fix bad Exception message

MoonBow-1 committed:
Remove return branching from JsonEventImpl.asEvent

MoonBow-1 committed:
Rename JsonEvent.asEvent() to JsonEvent.asEventMessage()

MoonBow-1 committed:
Rename JsonEvent.asNode() to JsonEvent.asPayloadJsonNode()

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEventImpl.java:
```diff
@@ -58,24 +58,32 @@ public JsonEventImpl(final JsonNode jsonNode) {
         // Event field completely missing
         if (!this.asNode().has("event")) {
-            return new EventMessageStub();
+            eventMessage = new EventMessageStub();
```
stub is immutable an therefore the same instance can and should be shared accross all stub returns, please change as such.

MoonBow-1 replied:
EventMessageStub instance is now shared from the EventManager to all JsonEventImpls, fixed in commit a24fded5

MoonBow-1 committed:
Use same instance of EventMessageStub for all Stub EventMessages

MoonBow-1 committed:
Make JsonEventImpl.asPayloadJsonNode return an empty Node if it's null or not an JSON object

- Does not throw an exception anymore

MoonBow-1 committed:
Rename JsonEvent.asTimeNode() tp JsonEvent.asTimeJsonNode()

MoonBow-1 committed:
Remove chained method calls in EventManager.convertData

MoonBow-1 committed:
Remove UUID from ExceptionJsonResponse.asJsonString message, and add an id field for it

MoonBow-1 committed:
Fix missing request field from ExceptionEvent

MoonBow-1 committed:
Log additional information from ExceptionEvent.logException

- Now uses ExceptionEventContext for the data
- toString() methods implemented for XForwarded* objects
- equals() and hashCode() methods implemented for Xforwarded*Stubs

MoonBow-1 committed:
Make JsonEventImpl.asPayloadJsonNode throw IllegalArgumentException instead of IllegalStateException

MoonBow-1 committed:
Make EventManager.convertData throw an Exception if parsing fails, and catch it in HECServiceImpl.sendEvents

MoonBow-1 committed:
Implement new integration test for failed requests and fix issues found there

- Response code can now be set by each Response type
- Integration Test tests that a failed request returns response code 400 and the correct response message and an id field
- Status and Application types can be removed from the Response implementations

MoonBow-1 commented:
I'll remove the HttpStatus fields from Responses tomorrow, and look into refactoring the `contentType()` method found in Response implementations, since now the application returns `ResponseEntity<JsonNode>` instead of String.

This is because of how Spring Boot handles response codes

MoonBow-1 committed:
Status and application methods removed from Response and added to ResponseEntities

- This has removed the status fields from all Response implementations
- ResponseEntities contain the status code and content type

MoonBow-1 committed:
Update Jmeter test plan

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEventImpl.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Override
+    public EventMessage asEventMessage() {
```
please make the method throw in case the content is bad, the stub is only useful for optional values

MoonBow-1 committed:
Throw an Exception instead of returning Stub from JsonEventImpl.asEventMessage

MoonBow-1 committed:
Add integration test for events that are integers

- These requests are supposed to fail

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEventImpl.java:
```diff
@@ -0,0 +1,122 @@
+
+    @Override
+    public JsonNode asTimeJsonNode() {
+        return this.jsonNode.get("time");
```
would 
```java
        if (!this.jsonNode.has("time")) {
            throw new EventFieldException("Time field is missing");
        }
```

be appropriate here?

MoonBow-1 replied:
Added and tested in commit [3ccdf30](https://github.com/teragrep/cfe_16/pull/47/commits/3ccdf308508a7f2eeed67dd0c959efd8f677cac2)

MoonBow-1 committed:
Add guard clause for JsonEvent.asTimeJsonNode

eemhu commented on src/main/java/com/teragrep/cfe_16/EventManager.java:
```diff
@@ -185,58 +181,12 @@ public ObjectNode convertData(
-        JsonNode jsonObject;
-        try {
-            jsonObject = this.objectMapper.readTree(eventInJson);
+            return new AcknowledgedJsonResponse("Success", ackId);
```
maybe here a variable for the return value could be used, so the method could only have one return statement at the end

MoonBow-1 replied:
Single return added in commit 661e6e32

eemhu commented on src/main/java/com/teragrep/cfe_16/response/AcknowledgedJsonResponse.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public ResponseEntity<JsonNode> asJsonNodeResponseEntity() {
+        final var jsonNode = new ObjectMapper().createObjectNode().put("message", body).put("ackID", ackID);
```
remove var keyword and replace with actual type

MoonBow-1 replied:
Actual type specified in commit acd93b96

eemhu commented on src/main/java/com/teragrep/cfe_16/response/ExceptionJsonResponse.java:
```diff
@@ -0,0 +1,89 @@
+
+    public ResponseEntity<JsonNode> asJsonNodeResponseEntity() {
+        final ObjectMapper jsonObjectBuilder = new ObjectMapper();
+        final var jsonNode = jsonObjectBuilder
```
change var to actual type

MoonBow-1 replied:
Actual type specified in commit 66342aaa

eemhu commented on src/main/java/com/teragrep/cfe_16/rest/HECRestController.java:
```diff
@@ -79,47 +80,48 @@ public class HECRestController {
     @RequestMapping(
             value = "services/collector",
             method = RequestMethod.POST,
             consumes = MediaType.APPLICATION_JSON_VALUE
```
should this also have value for `produces`?

MoonBow-1 replied:
MediaType specified in commit c4428e6f

eemhu commented on src/main/java/com/teragrep/cfe_16/service/HECServiceImpl.java:
```diff
@@ -174,10 +182,20 @@ public ObjectNode sendEvents(HttpServletRequest request, String channel, String
-
-        return ackNode;
+        try {
+            return this.eventManager.convertData(authToken, channel, eventInJson, headerInfo, this.acknowledgements);
```
should change to use single return statement

MoonBow-1 replied:
Single return added in commit 198283e7

eemhu commented:
otherwise ok

MoonBow-1 committed:
Use single return statement for EventManager.convertData

MoonBow-1 committed:
Use single return statement for HECServiceImpl.sendEvents

MoonBow-1 committed:
Specify MediaType for all HECRestController.sendEvents

MoonBow-1 committed:
Use exact object type for a variable in AcknowledgedJsonResponse.asJsonNodeResponseEntity

MoonBow-1 committed:
Use exact object type for a variable in ExceptionJsonResponse.asJsonNodeResponseEntity
